web/src/message_list_data_cache.ts

```diff
@@ -0,0 +1,33 @@
+import type {Filter} from "./filter";
+import type {MessageListData} from "./message_list_data";
+
+const cache: MessageListData[] = [];
+
+export function get(filter: Filter): MessageListData | undefined {
+    for (const cached_data of cache) {
+        if (cached_data.filter.equals(filter)) {
+            return cached_data;
+        }
+    }
+    return undefined;
+}
+
+export function add(message_list_data: MessageListData): void {
+    for (let i = 0; i < cache.length; i += 1) {
+        // Replace existing data in the cache if the filter is the same.
+        if (cache[i]?.filter.equals(message_list_data.filter)) {
+            cache[i] = message_list_data;
+            return;
+        }
+    }
+
+    cache.push(message_list_data);
+}
```

Comment:
We can leave a TODO comment, but probably we want it to pop the oldest that isn't `all_messages_data` if there are more than some constant, say 100, to avoid potential for infinite memory leaks.

Reply:
Makes sense, added a TODO comment.

Reply:
I don't think we can merge this without addressing the memory leak concern; the failure modes it can create and weird and hard to debug.

Reply:
Updated the code to implement that.